### PR TITLE
kast parser

### PR DESCRIPTION
**Pull request by liyili2**

### CODE REVIEW

**Comment by liyili2** [comment]
@dwightguth , I am fixing the bug for my style in my code. However, I don't understand something really. I have the following two style bugs:

[checkstyle] /var/lib/jenkins/jobs/k-framework-pull-requests/workspace/k/src/javasources/KTool/src/org/kframework/parser/kast/KastParserTokenManager.java:892: File has non-Unix line endings or trailing whitespace
[checkstyle] /var/lib/jenkins/jobs/k-framework-pull-requests/workspace/k/src/javasources/KTool/src/org/kframework/parser/kast/SimpleCharStream.java:77:5: Catching 'Throwable' is not allowed.

However, the two java files are all generated from my KastParser.jj file. I even don't understand why there is style problem in these two files.

**Comment by yilongli** [comment]
You don't need to fix the style in files generated by the JavaCC because they shouldn't be committed to the git repository.

**Comment by liyili2** [comment]
Good with auto-generated kast files now.

**Comment by grosu** [comment]
Can you avoid sending so cryptic messages?  What do you mean?  You should NOT include auto-generated files to the repo, that was Yilong's point.

Grigore

**Comment by liyili2** [comment]
If you go to parser/kast directory and look carefully, you will find that I never include auto-generated files in the repo. I am surprised that you also don't know that the style check is applied after we generate all files from JavaCC parser. For example, the basic parser will generate all auto-generated files before the style check is applied. The way to handle this is to add one more line in build.xml file to exclude all Java files in parser/kast directory when we are checking styles of files, which was the thing I wanted to ask and I figure out by myself now. That is why I said "Good now".

**Comment by grosu** [comment]
No, it is not good.  The one line that you mentioned needs to be changed in the script and your "good for now" answer was bad.  Please add a task for whoever is in charge of the script and make an effort to avoid confusing comments.

Grigore

Sent from phone

-------- Original message --------
From: Liyi Li
Date:08/08/2014 08:31 (GMT-06:00)
To: kframework/k
Cc: "Rosu, Grigore"
Subject: Re: [k] kast parser (#807)

If you go to parser/kast directory and look carefully, you will find that I never include auto-generated files in the repo. I am surprised that you also don't know that the style check is applied after we generate all files from JavaCC parser. For example, the basic parser will generate all auto-generated files before the style check is applied. The way to handle this is to add one more line in build.xml file to exclude all Java files in parser/kast directory when we are checking styles of files, which was the thing I wanted to ask and I figure out by myself now. That is why I said "Good now".

File: src/javasources/KTool/src/org/kframework/kil/KSorts.java [hunk 1]
```diff
@@ -26,4 +27,4 @@
     public static final String MAP_ITEM = "MapItem";
     public static final String SET = "Set";
     public static final String SET_ITEM = "SetItem";
-}
+}
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/kil/KSorts.java:31]: I should probably make a checkstyle check for this, but you should have the newline at the end of the file. It's standard for text files.

**Comment by dwightguth** [comment]
@grosu I need you to look at the grammar in the .jj file and let me know if it is okay with you. It doesn't seem to be completely in sync with the kast wiki page.

**Comment by grosu** [comment]
I know.  My plan was to read it as I define the syntax of kast in K, but that may take me until Monday.

Since this grammar is not really used by our parser, one other possibility is to just let it go through and then we fix it as we nail down the kast syntax for good.  If that does not bother you much, it is OK with me to go with it.

File: src/javasources/KTool/test/org/kframework/parser/kast/KastParserTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import org.kframework.kil.Term;
+import org.kframework.kil.loader.Context;
+
+public class KastParserTest {
+
+    @Test
+    public void test() {
+        final Context context = new Context();
+        //test input the .Bag
+        assertTrue(KastParser.parse("output.txt", ".Bag", context).getBody().equals(Bag.EMPTY));
```
Comment (dwightguth) [inline src/javasources/KTool/test/org/kframework/parser/kast/KastParserTest.java]: assertEquals

File: src/javasources/KTool/test/org/kframework/parser/kast/KastParserTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        //test input the number 0 and the number with ktoken format
+        assertTrue(KastParser.parse("output.txt", "0", context).getBody()
```
Comment (dwightguth) [inline src/javasources/KTool/test/org/kframework/parser/kast/KastParserTest.java]: it is bad practice to write a test that compares two values generated by the class under test against each other. Compare values from the class under text against statically constructed expected values.

File: src/javasources/KTool/src/org/kframework/parser/kast/KastParser.jj [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+Term parseKApp(Context context)
+: {Term a=null; Term b=null;Token n=null;Token m=null;}
+{".K" {return KSequence.EMPTY;}
+| "HOLE" {return Hole.KITEM_HOLE;}
+| n= < LABELCONSTANT >
+    (< LPR > b=parseKList(context) < RPR >)?
+  {if(b!=null)
+      return new KApp(KLabelConstant.of(n.image.substring(1,n.image.length()-1)),b);
+   else
+      {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/parser/kast/KastParser.jj:280]: I am not at all satisfied about how you are getting the labels from the image of the tokens here. There is a lot of hacky string manipulation going on here. Find a way to rewrite the grammar or else create a centralized helper method which clearly converts back and forth.

File: src/javasources/KTool/test/org/kframework/parser/kast/KastParserTest.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        //test parsing klabel and cells
+        Cell testCell = new Cell("abc",Bag.EMPTY);
+        assertTrue(KastParser.parse("output.txt", "<abc> .Bag </abc>", context)
+                    .getBody().equals(testCell));
+        //test parsing a KItem
+        List<Term> testKList = new ArrayList<Term>();
+        testKList.add(IntBuiltin.of(0));
+        testKList.add(IntBuiltin.of(0));
+        assertTrue(KastParser.parse("output.txt", "`abc`(0,0)", context)
+                .getBody().equals(new KApp(KLabelConstant.of("abc"), (Term)(new KList(testKList)))));
```
Comment (dwightguth) [inline src/javasources/KTool/test/org/kframework/parser/kast/KastParserTest.java]: KApp.of("abc", IntBuiltin.of(0), IntBuiltin.of(0))

File: src/javasources/KTool/src/org/kframework/parser/kast/KastParser.jj [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+  " "
+| "\r"
+| "\t"
+| "\n"
+}
+
+TOKEN : /* klabels */
+{ < NUM : ("-")?(< DIGIT >)+ >
+| < #DIGIT : [ "0"-"9" ] >
+| <STRING: "\"" (~["\"", "\n"] |
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/parser/kast/KastParser.jj:107]: These should not be hard coded into this parser. Tokens should have a clear syntax that allows encoding of arbitrary strings, and then the K tool determines whether they are an Int, Bool, String, etc.

Comment (liyili2) [inline src/javasources/KTool/src/org/kframework/parser/kast/KastParser.jj:107]: I understand what you are trying to say, but I don't know how to do this in JavaCC. I am not sure if you can do this in JavaCC.

Comment (radumereuta) [inline src/javasources/KTool/src/org/kframework/parser/kast/KastParser.jj:107]: But you do need strings for cell properties for example.

Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/parser/kast/KastParser.jj:107]: For cell properties yes but not for klabels or ktokens.

Comment (liyili2) [inline src/javasources/KTool/src/org/kframework/parser/kast/KastParser.jj:107]: So what is your suggestion, dwight? How to change it.

Comment (liyili2) [inline src/javasources/KTool/src/org/kframework/parser/kast/KastParser.jj:107]: Actually, I am a little confused what are you suggesting. However, The token in JavaCC acts like a lexer in a parser. If I have defined String = ""\"" (~["\"", "\n"] | "\\"" | "\n" | "\r" | "\t")\* "\"", then every strings you input will be recognized as String entities. If you want to define other things which are strings, such as "true". Most likely, the JavaCC is not able to recognize them as the precise things you want in the lexical analysis stage. They will be only recognized as String.

**Comment by dwightguth** [comment]
Closing. Please reopen on master

**Pull request closed without merging**